### ISSUE TITLE
sd-import chops off query part of URL

### ISSUE THREAD

**Issue by septatrix** [systemd/systemd#30696]
### systemd version the issue has been seen with

255

### CPU architectures issue was seen on

None

### Component

systemd-import

### Expected behaviour you didn't see

URLs from e.g. sysupdate.d files should retain their query part when fetching content

### Unexpected behaviour you saw

The query part is dropped completely. There relevant code can be found here:

https://github.com/systemd/systemd/blob/e40a67809b8a5a6988ce094de37c06c6f6b3afcf/src/shared/import-util.c#L103

### Steps to reproduce the problem

1. Create sysupdate file
    
    ```ini
    # test.conf
    [Source]
    Type=url-file
    Path=http://localhost:8000/?foo=bar
    MatchPattern=foobar_@v
    
    [Target]
    Path=/tmp
    ```

2. Open netcat: `nc -l -p 8000`
3. Run systemd-sysupdate: `SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/systemd-sysupdate --definitions . list`

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Netcat:

GET /SHA256SUMS HTTP/1.1
Host: localhost:8000
User-Agent: systemd-pull/253.14-1.fc38
Accept: */*


sysupdate:

Discovering installed instances…
Discovering available instances…
Manifest web cache miss for http://localhost:8000/?foo=bar.
⤵ Acquiring manifest file http://localhost:8000/SHA256SUMS…
Successfully forked off '(sd-pull)' as PID 36294.
Skipping PR_SET_MM, as we don't have privileges.
Writing downloaded data to STDOUT.
Pulling 'http://localhost:8000/SHA256SUMS'.
*   Trying 127.0.0.1:8000...
* Found bundle for host: 0x555d9dcb2d30 [serially]
* Server doesn't support multiplex (yet)
* Connection #0 is still name resolving, can't reuse
* Hostname localhost was found in DNS cache
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /SHA256SUMS HTTP/1.1
Host: localhost:8000
User-Agent: systemd-pull/253.14-1.fc38
Accept: */*
```

**Comment by poettering** [comment]
what did you expect instead? that we retain the `?` suffix, and insert the SHA256SUMS string in between?

**Comment by septatrix** [comment]
> what did you expect instead? that we retain the `?` suffix, and insert the SHA256SUMS string in between?

Yes, exactly

**Comment by septatrix** [comment]
Yes, technically they are per resource though in the context where they are used in systemd (esp. sd-sysupdate) one might consider all constructed URLs to be part of the same (re)`Source=`. Thus they would be more akin to defaults which are to be applied to every URL.

My two motivation are 1) metrics about the fleet status, e.g. by adding the current image id as a query parameter and then counting the share of each version, and 2) supporting mechanisms which are not (yet) possible with sd-update, or quite cumbersome. E.g. a server-controller slow rollout could be implemented by returning different SHA256SUMS based on the machine-id or hostname when they are transmitted as part of the query params. Another scenario would be appending some SoC info as query params and filtering the SHA256SUMS based on it (relevant for bootloader/device tree stuff which not only depends on Arch but also Chip). Currently one has to use different directories can lead to file duplication.
(This obviously does not work with static file servers but simple edge functions or even complex Apache rewrite filter could handle these.)

**Comment by septatrix** [comment]
> but would that be right? I mean, ? denotes parameters for the specified resource. But if we change the filename, we reference a different resource, hence does it make sense to still pass the parameters?

Alternatively we could add some way to specify headers though I do not thing sd-pull currently supports that. In that case I would be sufficient to simply have a global `pull.conf` which specifies headers. Even simply being able to set `User-Agent` would be enough for my purpose.
